### PR TITLE
[5.7] Add Query/Builder relationships

### PR DESCRIPTION
**Pull request by Calytic**
1. I have often encountered times where I need a relationship but the built-in relationships don't support my particular case and I must rewrite my relationship as a regular query/builder 
2. I don't want to write a custom relationship class
3. If I use a query/builder then I can't take advantage of the existing fantastic relationship magic

I've implemented this for my own projects, I'm halfway thinking it might be an enhancement so I'm passing it along.

### CODE REVIEW

**Comment by sisve** [comment]
Could you add some tests for these new scenarios?

**Comment by taylorotwell** [comment]
Agree with @sisve we would want some tests here.

**Comment by deleugpn** [comment]
I'm quite interested in better understanding how this works. Use cases and custom relationship configs, etc.
Could be a pretty strong selling point of 5.7

**Comment by Calytic** [comment]
I don't have a laravel dev environment setup and it would take much longer for me to familiarize myself with your testing regimen than it would someone already familiar with it to whip up some tests.  Sorry for that, I hope the code and information I provided may be of some use to your fantastic development team!

**Comment by crynobone** [comment]
A tests is not something you should ignore if you want this to be approved. If no one understands the full usage of this code and continuously run tests against it there are high chances that someone might introduce a failing bug or even removed it.

**Comment by Calytic** [comment]
I won't argue with you about the necessity of testing.  I'm just a lowly application developer on multiple deadlines at the moment. TDD is not part of my normal workflow (but I appreciate it's importance), and I have only done TDD on very small projects with significantly smaller scope and complexity.  While I relish an opportunity to contribute in that way, the uptick for me would be substantial and with much disappointment must decline at this time.

On a side note, I could see a scenario where the spirit of this pull request is accomplished in a different way, say with a new relationship type, rather than baking it in via `getRelationshipFromMethod` as I did.  

All that said, I will provide an example use-case and explain my reasoning for the change, if that helps.

I have, for example, a model `ResourceModel` with a bitwise enum `category_bitwise` field.

The `category_bitwise` field also has a `CategoryModel` associated with it, but unlike a normal model, the primary keys are bitwise, meaning the ID's are like this..

```
ID Title
1 "Category 1"
2 "Category 2"
4 "Category 3"
8 "Category 4"
16 "Category 5"
32 "Category 6"
etc..
```

I'm not particularly interested in a debate about whether or not this is how I should do it (it's just an example) and let's just say for the sake of discussion..

1. I have weighed the pros and cons and have determined that this approach of categorizing a `ResourceModel` better fits my use-case (as opposed to using a pivot table).  
2. The category ID's are populated from another system that I don't have control over and I'm using the bitwise strategy to maintain parity with that system.  
3. It's a performance concern and I need to reduce the number of database queries.

Now in my `ResourceModel` I want to create a relationship to `CategoryModel`, which I would normally do like this..

```php
class ResourceModel extends Model {
    public function category() {
        return $this->hasOne(CategoryModel::class,'id','category_bitwise')
    }
}
```

This approach, however, means that though my `CategoryModel` uses a bitwise enum, and thus should theoretically support resources being assigned multiple categories, it can only be assigned one category.

Having determined that I do not want or need to use a pivot table, and taking the path of least resistance, I rewrite my method to use a query:

```php
class ResourceModel extends Model {
    public function categories() {
        $IDs = getCategoryIDsFromBitwise($this->category_bitwise);
        return CategoryModel::whereIn('id', $IDs);
    }
}
```

As an aside and for the sake of clarity, `getCategoryIDsFromBitwise` performs a bitwise operation on `category_bitwise` and derives an array of individual ID's for 'CategoryModel' like so..

```php
$result = getCategoryIDsFromBitwise( 6 );
print_r($result); // [ 0 => 2, 1 => 4 ]
```

More detail on a bitwise enumeration implementation in php can be found here: https://stackoverflow.com/questions/254514/php-and-enumerations
or doing a quick search in your preferred search engine.

Now, I have rewritten my relationship as a query and am unfortunately left with the realization that I cannot use the built-in relationship magic provided by Laravel.  Most importantly, I cannot

1. Retrieve the categories using getAttribute via `$resource->categories`, instead having to do `$resource->categories()->get()`
2. Rely on Laravel to cache the results of my relationship, as it would normally via `setRelation`.

Then I have to one of the following things:
1. Override __get and implement my own caching
2. Use a different pattern throughout my application when interacting with the `categories` relationship
3. Implement a custom relationship class which effectively reduces to doing the exact same query with considerably more boilerplate

All that considered, I proposed this change for the sake preserving the ability to use the relationship magic on raw query/builder type relationships while preventing the need to do so with onerous boilerplate or using a unconventional pattern throughout my application.

Thank you for your time and consideration, and again I hope that this has been helpful despite my admittedly disappointing limitation regarding proper test implementation.

File: src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php [hunk 1]
```diff
@@ -6,7 +6,9 @@
 use DateTimeInterface;
 use Illuminate\Support\Arr;
 use Illuminate\Support\Str;
+use Illuminate\Database\Query;
```
Comment (dmitrybubyakin) [inline src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]: Does this class exist?

Comment (Calytic) [reply src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]: Nope, thanks for it pointing.

**Comment by deleugpn** [comment]
I would be willing to contribute with tests for this, but your post only defends your idea and doesn't show how you actually interact with the Query Builder relation.
Could you provide a bit of code on how this change makes your code simple to use the relationship via the builder?

**Comment by sisve** [comment]
The described scenario is a one-to-many-ish relation (stored as a bitmask). Can we add support for modifying the relation in this case; like adding/removing categories? Would we need a custom one-to-many relation class with abstract methods (add, remove, get) the user needs to implement?

**Comment by Calytic** [comment]
Sure,

Building on the previous examples..

```php
class ResourceModel extends Model {
    public function categories() {
        return CategoryModel::whereIn('id',  [2, 4]);
    }

    public function hasCategory($id) {
        foreach($this->categories as $category) {
            if($category->id == $id) {
                return true;
            }
        }
        return false;
    }
}
```

```php
class CategoryModel extends Model {

}
```

```php
$category1 = new CategoryModel(['id' => 2]);
$category2 = new CategoryModel(['id' => 4]);
```

```php
$resource = new ResourceModel;
assert(count($resource->categories) == 2); // true
assert($resource->hasCategory(2)); // true
assert($resource->hasCategory(4)); // true
assert( ! $resource->hasCategory(6)); // true
```

Hope this helps..

**Comment by deleugpn** [comment]
hey @Calytic 

Sorry for the delay, I was having a hard time understanding the concept of this. Today I got a spare time and worked on some quick tests. I opened a Pull Request on your fork (https://github.com/Calytic/framework/pull/1).

Unfortunately, I dont' bring good news. Obviously it's Taylor's call, but I don't think this fit nicely in the core.

- Eager Loading does not work. The relationship classes offer `addEagerConstraints` and `getRelated` methods which the Builder does not offer. That means we cannot do things like `User::with('posts')->get();` if `posts` is a Builder relationship.

- Count Eager Loading does not work. Same as the previous one. `User::withCount('posts')` does not work for lack of `getRelated` method.

- When using Eloquent Builder, we get an Eloquent object back, which kind of plays nicely with Eloquent relationships. However, when using Query Builder, we get an StdClass back and breaks every feature on Eloquent.

I didn't have enough time to go over all the use-cases, but the reason I'm voting down on this implementation is because of user expectation. As a Laravel user, I expect relationships to work in a certain way and if I ever need to reach for Query Builder or Eloquent Builder as a relationship, all those expectations would not be met. It would require an extensive documentation explaining these limitations and I don't think it's that much worth right now.

I think overall the idea is incredible and makes a lot of sense. However, it might require some major rework on Eloquent in order to make it really play nice. By that I mean refactoring existing relationships and trying to extract a first-class citizen to be able to handle Eloquent, Query Builder and Eloquent Builder relationships through a common interface.

**Comment by taylorotwell** [comment]
No plans on adding this currently.

**Pull request closed without merging**